data/code/lib.c

```diff
@@ -589,3 +593,793 @@ void check_lap_ex()
 #endif
 }
 #endif
+
+// This is in its own section despite not needing to be for now as it will have
+// to handle cases where one custom material is enabled and another is not.
+#if EXTENDED_TERRAIN_TYPES
+#if !GM4E01_DEBUG_BUILD
+// Check against all of the custom material flags enabled by the patch.
+int is_terrain_type_mod(char* terrain_type)
+{
+    int ret = 0;
+    if (*terrain_type == 0xB0)
+    {
+        ret = 1;
+    }
+    return ret;
+}
+
+// This and skip_item are separated to make bug reports easier to diagnose in the future.
+int should_return_fake_code(char* const ground)
+{
+    int* ground_materials = (int*)(ground + 0x20);  // Pointer to the triangle's material in memory.
+    int ret = 0;
+
+    if (*ground_materials != 0)
+    {
+        char* terrain_type = (char*)(*ground_materials + 0x16);  // Pointer to the collision flag.
+        ret = is_terrain_type_mod(terrain_type);
+    }
+
+    return ret;
+}
+
+int should_skip_item_inval(char* ground)
+{
+    int* ground_materials = (int*)(ground + 0x20);
+    int ret = 0;
+
+    if (*ground_materials != 0)
+    {
+        char* terrain_type = (char*)(*ground_materials + 0x16);
+        ret = is_terrain_type_mod(terrain_type);
+    }
+
+    return ret;
+}
+
+// Game will search for a Splash object due to the material hash being used.
+// This nullifies that behaviour.
+void get_splash_id_inline()
+{
+    register char* const ground asm("r3");  // R3 is a CrsGround object.
+
+    if (should_return_fake_code(ground) == 1)
+    {
+        asm("lis %r3, 0x00000");
+        asm("ori %r3, %r3, 0x0000");
+    }
+    else
+    {
+        asm("lwz %r3,0x20(%r5)");  // Original instruction.
+    }
+}
+
+void get_splash_height_inline()
+{
+    get_splash_id_inline();
+}
```

Comment:
Out of curiosity, what's the deal with this? Could we modify the `branchlink()` calls to jump to the same symbol if that's what the game requires? (The symbol would be given a name that combines the name of both callers.)

Reply:
That's my bad. This interferes a small piece of code that was inlined from a function call in the Debug version of the game, `CrsGround::getSplashCode`. It does require two hooks from the places where the function was inlined in the retail versions, but I did not consider that the `branchlink()` could just have the same function written in twice instead of what I did here.